integration/spadev0/main.go

```diff
@@ -0,0 +1,200 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"net/http"
+	"os"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	_ "github.com/joho/godotenv/autoload"
+	"github.com/klauspost/compress/zstd"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+)
+
+var logger = logging.Logger("spade-v0-tasks")
+
+func main() {
+	app := &cli.App{
+		Name:  "spadev0",
+		Usage: "run spade v0 CID sampling task",
+		Flags: []cli.Flag{
+			&cli.StringSliceFlag{
+				Name:        "sources",
+				DefaultText: "http://src-1/replicas.json.zst,http://src-2/replicas.json.zst",
+				Usage:       "comma-separated list of sources to fetch replica list from",
+				Required:    true,
+			},
+		},
+		Action: func(cctx *cli.Context) error {
+			ctx := cctx.Context
+
+			// logging.SetLogLevel("spade-v0-tasks", "DEBUG")
+
+			// Extract the sources from the flag
+			sources := cctx.StringSlice("sources")
+
+			for _, source := range sources {
+				res, err := fetchActiveReplicas(ctx, source)
+
+				if err != nil {
+					return err
+				}
+
+				var perProvider = make(map[int]ProviderReplicas)
+
+				for _, replica := range res.ActiveReplicas {
+					for _, contract := range replica.Contracts {
+						providerID := contract.ProviderID
+						size := 2 << replica.PieceLog2Size % 30 // Convert to GiB
```

Comment:
Two errors on this line, review the rest of the code regarding these ( bit math is weeeeird 😅 )

This `% 30` is not valid here (and anywhere really), as if the PieceLog2Size is 512M: it would remain so, whereas you wanted 0 or 1 depending on what you do with the exponent later. You need to change it to `(log2size - 30)`

Also the `log2` is misapplied: the value for e.g. 512M is 29. `2<<29` gives you 1GiB, you want `1<<x` to "reverse a log2" properly.

Reply:
Actually I *also* wrote an error here, what you *really* want is:
`(1<<log2size)>>30`